### ISSUE TITLE
opt: push filters into set operations

### ISSUE THREAD

**Issue by justinj** [cockroachdb/cockroach#35716]
We should push filters into set operations, this currently requires a pair of table scans:
```
root@127.0.0.1:51399/defaultdb> create table x (a int primary key);
CREATE TABLE

Time: 6.269ms

root@127.0.0.1:51399/defaultdb> create table y (a int primary key);
CREATE TABLE

Time: 9.03ms

root@127.0.0.1:51399/defaultdb> explain (opt) select * from (select * from x union all select * from y) where a = 3;
        text
+------------------+
  select
   ├── union-all
   │    ├── scan x
   │    └── scan y
   └── filters
        └── a = 3
```

This might be a good starter project.

**Comment by khalibartan** [comment]
@justinj I would like to work on this one. Can you elaborate a bit more to help me getting started on this one?

**Comment by justinj** [comment]
Hey @khalibartan, this was actually fixed by #37418, though this issue was not closed. Sorry about the confusion!
